File: app/modules/planning/domain/calculations/opportunities/profit_taking.py
"""Profit-taking opportunity calculator.

Identifies windfall positions that should be trimmed for profit-taking.
"""

from typing import Any, Dict, List, Optional

from app.domain.services.exchange_rate_service import ExchangeRateService
from app.domain.value_objects.trade_side import TradeSide
from app.modules.planning.domain.calculations.opportunities.base import (
    OpportunityCalculator,
    OpportunityContext,
    opportunity_calculator_registry,
)
from app.modules.planning.domain.models import ActionCandidate
from app.modules.scoring.domain.windfall import get_windfall_recommendation


class ProfitTakingCalculator(OpportunityCalculator):
    """Identifies positions with windfall gains to sell."""

    def __init__(self, exchange_rate_service: Optional[ExchangeRateService] = None):
        self.exchange_rate_service = exchange_rate_service

    @property
    def name(self) -> str:
        return "profit_taking"

    def default_params(self) -> Dict[str, Any]:
        return {
            "windfall_threshold": 0.30,  # 30% excess gain
            "priority_weight": 1.2,  # Prioritize profit-taking
        }

    async def calculate(
        self,
        context: OpportunityContext,
        params: Dict[str, Any],
    ) -> List[ActionCandidate]:
        """
        Identify profit-taking opportunities (windfall positions to trim).

        Checks each position for windfall gains (exceeding expected returns)
        and suggests selling a percentage to lock in profits.

        Args:
            context: Portfolio context with positions and securities
            params: Calculator parameters (windfall_threshold, priority_weight)

        Returns:
            List of sell ActionCandidates for windfall positions
        """
        opportunities = []

        for pos in context.positions:
            security = context.stocks_by_symbol.get(pos.symbol)
            if not security or not security.allow_sell:
                continue

            position_value = pos.market_value_eur or 0
            if position_value <= 0:
                continue

            # Check for windfall
            windfall_rec = await get_windfall_recommendation(
                symbol=pos.symbol,
                current_price=pos.current_price or pos.avg_price,
                avg_price=pos.avg_price,
                first_bought_at=(
                    pos.first_bought_at if hasattr(pos, "first_bought_at") else None
                ),
            )

            if windfall_rec.get("recommendation", {}).get("take_profits"):
                rec = windfall_rec["recommendation"]
                sell_pct = rec["suggested_sell_pct"] / 100
                sell_qty = int(pos.quantity * sell_pct)
                sell_value = sell_qty * (pos.current_price or pos.avg_price)

                # Convert to EUR
                exchange_rate = 1.0
                if pos.currency and pos.currency != "EUR":
                    if self.exchange_rate_service:
                        exchange_rate = await self.exchange_rate_service.get_rate(
                            pos.currency, "EUR"
                        )
                    else:
                        exchange_rate = 1.0  # Fallback
                sell_value_eur = (
                    sell_value / exchange_rate if exchange_rate > 0 else sell_value
                )

                # Apply priority multiplier inversely: higher multiplier = lower sell priority
                base_priority = windfall_rec.get("windfall_score", 0.5) + 0.5
                multiplier = params.get("priority_weight", 1.2)
                security_multiplier = security.priority_multiplier if security else 1.0
                final_priority = (base_priority * multiplier) / security_multiplier

                opportunities.append(
                    ActionCandidate(
                        side=TradeSide.SELL,
                        symbol=pos.symbol,
                        name=security.name,
                        quantity=sell_qty,
                        price=pos.current_price or pos.avg_price,
                        value_eur=sell_value_eur,
                        currency=pos.currency or "EUR",
                        priority=final_priority,
                        reason=rec["reason"],
                        tags=["windfall", "profit_taking"],
                    )
                )

        return opportunities


# Auto-register this calculator
_profit_taking_calculator = ProfitTakingCalculator()
opportunity_calculator_registry.register(
    _profit_taking_calculator.name, _profit_taking_calculator
)

File: trader-go/internal/modules/opportunities/calculators/profit_taking.go
package calculators

import (
	"fmt"

	"github.com/aristath/arduino-trader/internal/modules/planning/domain"
	"github.com/rs/zerolog"
)

// ProfitTakingCalculator identifies opportunities to take profits from positions with gains.
type ProfitTakingCalculator struct {
	*BaseCalculator
}

// NewProfitTakingCalculator creates a new profit taking calculator.
func NewProfitTakingCalculator(log zerolog.Logger) *ProfitTakingCalculator {
	return &ProfitTakingCalculator{
		BaseCalculator: NewBaseCalculator(log, "profit_taking"),
	}
}

// Name returns the calculator name.
func (c *ProfitTakingCalculator) Name() string {
	return "profit_taking"
}

// Category returns the opportunity category.
func (c *ProfitTakingCalculator) Category() domain.OpportunityCategory {
	return domain.OpportunityCategoryProfitTaking
}

// Calculate identifies profit-taking opportunities.
func (c *ProfitTakingCalculator) Calculate(
	ctx *domain.OpportunityContext,
	params map[string]interface{},
) ([]domain.ActionCandidate, error) {
	// Parameters with defaults
	minGainThreshold := GetFloatParam(params, "min_gain_threshold", 0.15)     // 15% minimum gain
	windfallThreshold := GetFloatParam(params, "windfall_threshold", 0.30)    // 30% for windfall
	minHoldDays := GetIntParam(params, "min_hold_days", 90)                   // Minimum holding period
	sellPercentage := GetFloatParam(params, "sell_percentage", 1.0)           // Sell 100% by default
	maxPositions := GetIntParam(params, "max_positions", 0)                   // 0 = unlimited

	if !ctx.AllowSell {
		c.log.Debug().Msg("Selling not allowed, skipping profit taking")
		return nil, nil
	}

	var candidates []domain.ActionCandidate

	c.log.Debug().
		Float64("min_gain_threshold", minGainThreshold).
		Float64("windfall_threshold", windfallThreshold).
		Int("min_hold_days", minHoldDays).
		Msg("Calculating profit-taking opportunities")

	for _, position := range ctx.Positions {
		// Skip if ineligible
		if ctx.IneligibleSymbols[position.Symbol] {
			continue
		}

		// Skip if recently sold
		if ctx.RecentlySold[position.Symbol] {
			continue
		}

		// Get current price
		currentPrice, ok := ctx.CurrentPrices[position.Symbol]
		if !ok || currentPrice <= 0 {
			c.log.Warn().
				Str("symbol", position.Symbol).
				Msg("No current price available")
			continue
		}

		// Calculate gain
		costBasis := position.AverageCost
		if costBasis <= 0 {
			continue
		}

		gainPercent := (currentPrice - costBasis) / costBasis

		// Check if gain meets threshold
		if gainPercent < minGainThreshold {
			continue
		}

		// Get security info
		security, ok := ctx.StocksBySymbol[position.Symbol]
		if !ok {
			continue
		}

		// Determine if windfall
		isWindfall := gainPercent >= windfallThreshold

		// Calculate quantity to sell
		quantity := position.Quantity
		if sellPercentage < 1.0 {
			quantity = int(float64(position.Quantity) * sellPercentage)
			if quantity == 0 {
				quantity = 1
			}
		}

		// Calculate value
		valueEUR := float64(quantity) * currentPrice

		// Apply transaction costs
		transactionCost := ctx.TransactionCostFixed + (valueEUR * ctx.TransactionCostPercent)
		netValueEUR := valueEUR - transactionCost

		// Calculate priority (higher gain = higher priority)
		priority := gainPercent

		// Windfall gets extra priority boost
		if isWindfall {
			priority *= 1.5
		}

		// Build reason
		reason := fmt.Sprintf("%.1f%% gain (cost basis: %.2f, current: %.2f)",
			gainPercent*100, costBasis, currentPrice)

		if isWindfall {
			reason = fmt.Sprintf("Windfall: %s", reason)
		}

		// Build tags
		tags := []string{"profit_taking"}
		if isWindfall {
			tags = append(tags, "windfall")
		}

		candidate := domain.ActionCandidate{
			Side:     "SELL",
			Symbol:   position.Symbol,
			Name:     security.Name,
			Quantity: quantity,
			Price:    currentPrice,
			ValueEUR: netValueEUR,
			Currency: string(security.Currency),
			Priority: priority,
			Reason:   reason,
			Tags:     tags,
		}

		candidates = append(candidates, candidate)
	}

	// Limit to max positions if specified
	if maxPositions > 0 && len(candidates) > maxPositions {
		// Sort by priority (descending) and take top N
		// For now, just truncate (sorting will be done by caller)
		candidates = candidates[:maxPositions]
	}

	c.log.Info().
		Int("candidates", len(candidates)).
		Msg("Profit-taking opportunities identified")

	return candidates, nil
}

func init() {
	// Auto-register on import
	// Will be replaced with actual logger when registry is initialized
	DefaultCalculatorRegistry.Register(NewProfitTakingCalculator(zerolog.Nop()))
}
